core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+    auto res = _chunkMap.insert(ele); // insert should fail if the key already exists.
+    if (!res.second) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTaskList::_insertChunkTask entry already existed " << chunkId);
+    }
+    return res.first;
+}
+
+/// @return true if this object is ready to provide a Task from its queue.
+bool ChunkTaskList::ready(bool useFlexibleLock) {
+    std::lock_guard<std::mutex> lock(_mapMx);
+    return _ready(useFlexibleLock);
+}
+
+/// Precondition: _queueMutex must be locked
+/// If true is returned, _readyChunk will point to a chunk with a Task that is ready to run.
+/// This function starts checking at the _activeChunk and only looks to the next chunk
+/// if there are no tasks to run on the current chunk.
+/// The _activeChunk advances when all of its Tasks have completed.
+/// @return true if this object is ready to provide a Task from its queue. _readyChunk != nullptr.
+bool ChunkTaskList::_ready(bool useFlexibleLock) {
+    if (empty()) {
+        return false;
+    }
+
+    // If the _activeChunk is invalid, start at the beginning.
+    if (_activeChunk == _chunkMap.end()) {
+        _activeChunk = _chunkMap.begin();
+        _activeChunk->second->setActive(); // Flag tasks on active so new Tasks added wont be run.
+    }
+
+    // Check the active chunk for valid Tasks
+    if (_activeChunk->second->ready(useFlexibleLock) == ChunkTasks::ReadyState::READY) {
+        _readyChunk = _activeChunk->second;
+        return true;
+    }
+
+    // Should the active chunk be advanced?
+    if (_activeChunk->second->readyToAdvance()) {
+        auto newActive = _activeChunk;
+        ++newActive;
+        if (newActive == _chunkMap.end()) {
+            newActive = _chunkMap.begin();
+        }
+
+        // Clean up the old _active chunk before moving on.
+        _activeChunk->second->setActive(false); // This should move pending Tasks to _activeTasks
+        // _inFlightTasks must be empty as readyToAdvance was true.
+        if (_activeChunk->second->empty()) {
+            if (newActive == _activeChunk) {
+                newActive = _chunkMap.end();
+            }
+            _chunkMap.erase(_activeChunk);
+        }
+
+        _activeChunk = newActive;
+        if (newActive == _chunkMap.end()) {
+            return false;
+        }
+        newActive->second->movePendingToActive();
+        newActive->second->setActive();
+    }
+
+    // Advance through chunks until READY or NO_RESOURCES found, or until entire list scanned.
+    auto iter = _activeChunk;
+    ChunkTasks::ReadyState chunkState = iter->second->ready(useFlexibleLock);
+    while (chunkState != ChunkTasks::ReadyState::READY
+           && chunkState != ChunkTasks::ReadyState::NO_RESOURCES) {
+        ++iter;
+        if (iter == _chunkMap.end()) {
+            iter = _chunkMap.begin();
+        }
+        if (iter == _activeChunk) {
+            return false;
+        }
+        chunkState = iter->second->ready(useFlexibleLock);
```

Comment:
Could we replace it by testing if _chunkMap size is equals 1? If yes this might be shorter?